### PR TITLE
[build-script] use explicit build/test list between build-script and build-script-impl

### PR DESCRIPTION
**Pull request by karwa**
#### What's in this pull request?

Instead of having lots of different skip-[platform] and skip-test-[platform] flags for every stdlib SDK, use an explicit list of targets to build and test.

There are no outward-facing changes (besides a more verbose and obvious configuration message). The build and test list are calculated from the same `--ios`, etc flags as they are today. This only affects how `build-script` tells `build-script-impl` which targets to build/test. It simplifies stdlib target calculation, and gives us the opportunity to override it with more granular options (e.g. just building one stdlib deployment target, not a whole SDK, or just building tests for one target).

Reopened from #2780 
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

(and appropriate test flags) with an explicit build and test list.

Made configuration message more visible and obvious

### CODE REVIEW

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please smoke test

**Comment by gribozavr** [comment]
Not a review, just triggering the CI.

File: utils/build-script [hunk 1]
```diff
@@ -1136,98 +1136,49 @@ details of the setups of other systems or automated environments.""")
...
+    # --validation-test implies --test.
+    if args.validation_test:
+        args.test = True
 
-    # --skip-{ios,tvos,watchos} or --skip-build-{ios,tvos,watchos} are
-    # merely shorthands for --skip-build-{**os}-{device,simulator}
-    if not args.ios or args.skip_build_ios:
+    # --long-test implies --test.
+    if args.long_test:
+        args.test = True
```
Comment (rintaro) [inline utils/build-script]: This is not true.
See 457f2b901caad14e2085ea0fe2c9b116b5df3520 for intent of `--long-test`

File: utils/build-script [hunk 2]
```diff
@@ -1338,10 +1289,61 @@ details of the setups of other systems or automated environments.""")
...
         build_script_impl_args += ["--skip-build-cmark",
                                    "--skip-build-llvm",
                                    "--skip-build-swift"]
+        args.skip_build_benchmarks = True
+        args.build_lldb = False
+        args.build_llbuild = False
+        args.build_swiftpm = False
+        args.build_xctest = False
+        args.build_foundation = False
+        args.build_libdispatch = False
```
Comment (rintaro) [inline utils/build-script]: Please do not mix `args` tweak and `build_script_impl_args` generation.
I really want to keep current main control flow:

1. [Parse arguments](https://github.com/apple/swift/blob/5959dd2aa1e9ab940ec478d6c88ab8d343c6a28b/utils/build-script#L971-L988)
2. [Validate and tweak `args`](https://github.com/apple/swift/blob/5959dd2aa1e9ab940ec478d6c88ab8d343c6a28b/utils/build-script#L995-L1206)
3. [Pre `build-script-impl` work](https://github.com/apple/swift/blob/5959dd2aa1e9ab940ec478d6c88ab8d343c6a28b/utils/build-script#L1212-L1250)
4. [Generate `build_script_impl_args` and call `build-script-impl`](https://github.com/apple/swift/blob/5959dd2aa1e9ab940ec478d6c88ab8d343c6a28b/utils/build-script#L1252-L1439)
6. [Post `build-script-impl` work](https://github.com/apple/swift/blob/5959dd2aa1e9ab940ec478d6c88ab8d343c6a28b/utils/build-script#L1441-L1457)

This control-flow makes Python migration easier.

File: utils/build-script [hunk 3]
```diff
@@ -1338,10 +1289,61 @@ details of the setups of other systems or automated environments.""")
...
+        skipped_targets += StdlibDeploymentTarget.AppleWatch.allArchs
+    if args.skip_build_watchos_simulator or not args.watchos:
+        skipped_targets += StdlibDeploymentTarget.AppleWatchSimulator.allArchs
+
+    build_stdlib_deployment_targets = [
+        x for x in args.stdlib_deployment_targets
+        if x not in skipped_targets]
+
+    if args.skip_build or len(build_stdlib_deployment_targets) == 0:
+        build_stdlib_deployment_targets = ['none']
```
Comment (rintaro) [inline utils/build-script]: Please consider to factor out the calculation of
`build_stdlib_deployment_targets` and `test_deployment_targets` into
`swift_build_support` package and write some tests.

**Comment by rintaro** [comment]
> This only affects how build-script tells build-script-impl which targets to build/test.

If so, you shouldn't have to modify [`args` tweak block](https://github.com/apple/swift/blob/5959dd2aa1e9ab940ec478d6c88ab8d343c6a28b/utils/build-script#L1044-L1206).
That said, I also think [these lines](https://github.com/apple/swift/blob/5959dd2aa1e9ab940ec478d6c88ab8d343c6a28b/utils/build-script#L1114-L1122) are redundant or even harmful.

File: utils/build-script (hunk 2 above)
Comment (ddunbar) [inline utils/build-script]: It might even be nice to break each of those phases into a separate function, and document the intent.

**Comment by karwa** [comment]
@rintaro You're right that I don't have to, but lots of the calculations there were redundant from a perspective of an explicit build list derived from a restricted list of configured targets, and then a test list derived from a restricted list of build targets.

Not only that, but it hid away some vital details, like the fact that iOS targets are sort configured backwards - rather than being configured and built by default, they are configured by default and skipped from building unless you add the `--ios` flag. I believe that's a lot clearer after the changes.

it may be possible to put some of the other argument-shuffling back, but actually breaking build/test list generation in to functions is probably better.

**Comment by ddunbar** [comment]
This and #2880 are going to conflict heavily, so we should probably figure out which makes sense to land first if we are going to go both directions.

**Comment by gribozavr** [comment]
I would suggest that we go with #2880.  It is more aggressive, but it should allow us to get rid of the shell script quicker.  Any concerns?

**Comment by karwa** [comment]
Shelved until build-script-impl is removed

**Pull request closed without merging**